### PR TITLE
Improve code coverage blockchain hook

### PR DESCRIPTION
**Pull request by mariusmihaic**
- Added tests for `BlockchainHook` to increase code coverage to **99,3%**
- The only paths that were not tests were:
1. A path for debugger in `stopMeasure`
2. `GetAllState` func which is dummy and shall be removed

### CODE REVIEW

File: process/smartContract/hooks/blockChainHook.go [hunk 1]
```diff
@@ -167,6 +167,10 @@ func checkForNil(args ArgBlockChainHook) error {
 
 // GetCode returns the code for the given account
 func (bh *BlockChainHookImpl) GetCode(account vmcommon.UserAccountHandler) []byte {
+	if check.IfNil(account) {
```
Comment (iulianpascalau) [inline process/smartContract/hooks/blockChainHook.go:170]: :+1:

File: process/smartContract/hooks/blockChainHook_test.go [hunk 2]
```diff
@@ -54,98 +61,185 @@ func createMockVMAccountsArguments() hooks.ArgBlockChainHook {
...
+		Function:      function,
+		RecipientAddr: receiver,
+		VMInput: vmcommon.VMInput{
+			CallerAddr: sender,
+		},
+	}
 }
 
-func TestNewBlockChainHookImpl_NilBlockChainShouldErr(t *testing.T) {
+func TestNewBlockChainHookImpl_NilAccountsAdapterShouldErr(t *testing.T) {
```
Comment (iulianpascalau) [inline process/smartContract/hooks/blockChainHook_test.go:74]: might rename the test just to `TestNewBlockChainHookImpl`

Comment (mariusmihaic) [reply process/smartContract/hooks/blockChainHook_test.go:74]: Right, missed it, thanks !

File: process/smartContract/hooks/blockChainHook.go (hunk 1 above)
Comment (mariusmihaic) [reply process/smartContract/hooks/blockChainHook.go:170]: :pray:

**Review by iulianpascalau — APPROVED** [review]

File: process/smartContract/hooks/blockChainHook_test.go [hunk 3]
```diff
@@ -583,10 +1480,40 @@ func TestBlockChainHookImpl_GetESDTToken(t *testing.T) {
...
+				require.Equal(t, invalidUnmarshalledData, buff)
+				return errMarshaller
+			},
+		}
+
+		bh, _ := hooks.NewBlockChainHookImpl(args)
+		bh.SetFlagOptimizeNFTStore(false)
+
+		esdtData, err := bh.GetESDTToken(address, token, nonce)
+		require.Nil(t, esdtData)
```
Comment (bogdan-rosianu) [inline process/smartContract/hooks/blockChainHook_test.go:1510]: not related to this PR, but we have some typos in this file. can you replace `addres` with `address` ?

File: process/smartContract/hooks/blockChainHook.go [hunk 4]
```diff
@@ -640,14 +640,10 @@ func (bh *BlockChainHookImpl) Close() error {
 func (bh *BlockChainHookImpl) ClearCompiledCodes() {
 	bh.compiledScPool.Clear()
 	err := bh.compiledScStorage.DestroyUnit()
-	if err != nil {
-		log.Error("blockchainHook ClearCompiledCodes DestroyUnit", "error", err)
-	}
+	log.LogIfError(err, "func", "BlockChainHookImpl.ClearCompiledCodes().compiledScStorage.DestroyUnit")
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/smartContract/hooks/blockChainHook.go:643]: IMHO, I don't feel like it changes the context. There is the same information, but wrapped in a different way.

File: process/smartContract/hooks/blockChainHook_test.go [hunk 5]
```diff
@@ -513,33 +828,615 @@ func TestBlockChainHookImpl_IsPayablePayableBySC(t *testing.T) {
...
+		found, actualCode := bh.GetCompiledCode(codeHash)
+		require.False(t, found)
+		require.Nil(t, actualCode)
+		require.False(t, wasCodeSavedInPool.IsSet())
+	})
+
+	t.Run("compiled code found in storage, but nil", func(t *testing.T) {
+		args := createMockBlockChainHookArgs()
+		args.NilCompiledSCStore = false
+		args.ConfigSCStorage = config.StorageConfig{
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/smartContract/hooks/blockChainHook_test.go:987]: Sure.
I've read this comment here:
```go
// LvlDB currently the only supported DBs
// More to be added
const (
	LvlDB       DBType = "LvlDB"
	LvlDBSerial DBType = "LvlDBSerial"
	MemoryDB    DBType = "MemoryDB"
)
```

And I thought that I'd stick with `LvlDB`, but it seems like this info is outdated. Refactored as your suggestion

File: process/smartContract/hooks/blockChainHook_test.go [hunk 6]
```diff
@@ -513,33 +828,615 @@ func TestBlockChainHookImpl_IsPayablePayableBySC(t *testing.T) {
...
+	t.Run("compiled code not found in compiled sc pool, get it from storage", func(t *testing.T) {
+		args := createMockBlockChainHookArgs()
+		args.ConfigSCStorage = config.StorageConfig{
+			Cache: config.CacheConfig{
+				Capacity: 10,
+				Type:     string(storageUnit.LRUCache),
+			},
+			DB: config.DBConfig{
+				FilePath:     "test2",
+				Type:         string(storageUnit.LvlDB),
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/smartContract/hooks/blockChainHook_test.go:1033]: Refactored it !

File: process/smartContract/hooks/blockChainHook_test.go [hunk 7]
```diff
@@ -583,10 +1480,40 @@ func TestBlockChainHookImpl_GetESDTToken(t *testing.T) {
 		require.NotNil(t, esdtData)
 		assert.Equal(t, emptyESDTData, esdtData)
 	})
+	t.Run("error unmarshall", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/smartContract/hooks/blockChainHook_test.go:1483]: Changed it. Interestingly, auto-correct did not detect it

File: process/smartContract/hooks/blockChainHook_test.go (hunk 3 above)
Comment (mariusmihaic) [reply process/smartContract/hooks/blockChainHook_test.go:1510]: Right, it should've been `addressContainer` actually. Refactoed it

File: process/smartContract/hooks/blockChainHook.go [hunk 8]
```diff
@@ -596,9 +600,7 @@ func (bh *BlockChainHookImpl) SetCurrentHeader(hdr data.HeaderHandler) {
 func (bh *BlockChainHookImpl) SaveCompiledCode(codeHash []byte, code []byte) {
 	bh.compiledScPool.Put(codeHash, code, len(code))
 	err := bh.compiledScStorage.Put(codeHash, code)
-	if err != nil {
-		log.Debug("SaveCompiledCode", "error", err, "codeHash", codeHash)
-	}
+	log.LogIfError(err, "codeHash", codeHash)
```
Comment (bogdan-rosianu) [inline process/smartContract/hooks/blockChainHook.go:603]: `log.LogIfError(err, "func", "BlockChainHookImpl.SaveCompiledCode: compiledScStorage.Put")`

File: process/smartContract/hooks/blockChainHook.go [hunk 9]
```diff
@@ -625,9 +627,7 @@ func (bh *BlockChainHookImpl) GetCompiledCode(codeHash []byte) (bool, []byte) {
 func (bh *BlockChainHookImpl) DeleteCompiledCode(codeHash []byte) {
 	bh.compiledScPool.Remove(codeHash)
 	err := bh.compiledScStorage.Remove(codeHash)
-	if err != nil {
-		log.Debug("DeleteCompiledCode", "error", err, "codeHash", codeHash)
-	}
+	log.LogIfError(err, "codeHash", codeHash)
```
Comment (bogdan-rosianu) [inline process/smartContract/hooks/blockChainHook.go:630]: `log.LogIfError(err, "func", "BlockChainHookImpl.DeleteCompiledCode: compiledScStorage.Remove")`

**Review by iulianpascalau — APPROVED** [review]

File: process/smartContract/hooks/blockChainHook.go (hunk 8 above)
Comment (mariusmihaic) [reply process/smartContract/hooks/blockChainHook.go:603]: Right. Missed it. Added it

File: process/smartContract/hooks/blockChainHook.go (hunk 9 above)
Comment (mariusmihaic) [reply process/smartContract/hooks/blockChainHook.go:630]: Right. Missed it. Added it

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]

**Review by sasurobert — APPROVED** [review]

**Pull request merged**